### PR TITLE
[swift] UnavailableFunction Swift rule

### PR DESCRIPTION
**Pull request by tomidelucca**
Before submitting a PR, please check that:
 - [ ] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**

We are adding a second Java rule for Swift. This rule checks that an availability modifier was added to any function that's unimplemented.

Here is an example:

```
class ViewController: UIViewController {
  public required init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
}
```
This is a perfectly valid piece of Swift code, but if you call the `init?(coder:)` initializer of the `ViewController` class the app will crash due to the `fatalError()`. This initializer **needs** to be implemented to extend `UIViewController`, so removing the method will not work.

This rule checks that you add the `@available(*, unavailable)` attribute to the function and initializer declarations. With this attribute present, the function will not be able to be called, but the compiler will see it as implemented so there won't be any compile errors.

The expected non-triggering version of the code snippet looks like this:
```
class ViewController: UIViewController {
  @available(*, unavailable)
  public required init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
}
```

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/rule/bestpractices/UnavailableFunctionRule.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        if (ctx == null || ctx.statements() == null) {
+            return false;
+        }
+
+        final List<SwiftParser.StatementContext> statements = ctx.statements().statement();
+
+        return statements.size() == 1 && FATAL_ERROR.equals(statements.get(0).getStart().getText());
+    }
+
+    private Boolean hasUnavailableModifier(final List<SwiftParser.AttributeContext> attributes) {
```
Comment (jsotuyod) [inline pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/rule/bestpractices/UnavailableFunctionRule.java]: do we need a boxed primitive here?

File: pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/rule/bestpractices/UnavailableFunctionRule.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            final SwiftParser.AttributesContext attributes = ctx.initializerHead().attributes();
+            if (attributes == null || !hasUnavailableModifier(attributes.attribute())) {
+                addViolation(data, ctx);
+            }
+        }
+
+        return null;
+    }
+
+    private Boolean shouldIncludeUnavailableModifier(final SwiftParser.CodeBlockContext ctx) {
```
Comment (jsotuyod) [inline pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/rule/bestpractices/UnavailableFunctionRule.java]: do we need a boxed primitive here?

File: pmd-swift/src/main/resources/category/swift/bestpractices.xml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    </rule>
+
+    <rule name="UnavailableFunction"
+          since="7.0"
+          message="Unimplemented functions should be marked as unavailable."
+          class="net.sourceforge.pmd.lang.swift.rule.bestpractices.UnavailableFunctionRule"
+          externalInfoUrl="http://pmd.github.io/pmd/pmd_rules_swift_bestpractices.html#unavailablefunction">
+        <description>
+            Due to Objective-C and Swift interoperability some functions are often required to be implemented but
+            aren't really needed. Is extremely common that the sole implementation of the functions consist of throwing
```
Comment (jsotuyod) [inline pmd-swift/src/main/resources/category/swift/bestpractices.xml]:
```suggestion
            aren't really needed. It is extremely common that the sole implementation of the functions consist of throwing
```

File: pmd-swift/src/main/resources/category/swift/bestpractices.xml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            a fatal error. Marking this functions as unavailable prevents them from being executed while still making
```
Comment (jsotuyod) [inline pmd-swift/src/main/resources/category/swift/bestpractices.xml]:
```suggestion
            a fatal error. Marking these functions as unavailable prevents them from being executed while still making
```

File: pmd-swift/src/main/resources/category/swift/bestpractices.xml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            the compiler happy.
+        </description>
+        <priority>3</priority>
+    </rule>
```
Comment (jsotuyod) [inline pmd-swift/src/main/resources/category/swift/bestpractices.xml]: please, provide an `<example>` with valid and invalid code

**Comment by jsotuyod** [comment]
@tomidelucca please rebase, #1877 has now been merged

**Comment by tomidelucca** [comment]
@jsotuyod done!

**Pull request merged**